<a href="https://colab.research.google.com/github/AlienX77-cmd/Python-for-Business-and-Finance/blob/main/Exercise_08.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Coding Exercises (Part 1)

Now, you will have the opportunity to practice what you have learned. <br>
__Follow the instructions__ and insert your code! 

The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need some help or if you want to check your code, you can also have a look at the __solutions__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Exercise 8:  Complex TVM and Capital Budgeting problems with Python and Numpy

In [2]:
# install numpy finance to google colab
pip install numpy-financial

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [3]:
# run the cell
import numpy as np
import numpy_financial as npf # NEW!!!

The ABC Company evaluates to buy an additional machine that will increase future profits/cashflows by
- 50 USD in t1, 
- 80 USD in t2,
- 120 USD in t3, 
- 150 USD in t4,
- 200 USD in t5. (each cf at period´s end)<br>

The machine costs 500 USD (Investment in to). ABC´s required rate of return (Cost of Capital) is 7.5% p.a. 

In [4]:
# run the cell
cf = np.array([-500, 50, 80, 120, 150, 200])
cf

array([-500,   50,   80,  120,  150,  200])

In [5]:
# run the cell!
r = 0.075

1. Calculate the Net __Present Value (NPV)__ of the Project with __numpy financial functions__.

In [7]:
NPV = npf.npv(r, cf)
NPV

-36.03468883756554

(result: -36.03468883756554)

2. Calculate the __IRR__ of the Project with __numpy financial functions__.

In [10]:
IRR = npf.irr(cf)
IRR

0.05234347404159134

(result: 0.05234347404159134)

3. Should ABC __accept or reject__ the project?

In [11]:
print("Reject because NPV < 0")

Reject because NPV < 0


4. Assume a 20-year-old investor wants to retire in 45 years at the age of 65. He expects to earn 4.5% p.a. on his Investment during the Funding Phase and during the Payout Phase. He plans to start depositing 500 USD at the end of each month for n years in order to be able to withdraw 3,000 USD per month (at the beginning of each month) for 30 years during retirement. The Annuity Contract´s final Value shall be 50,000 USD when he is 95 years old (to cover some more years). __In how many months will the investor´s funding phase start__?

__Payout Phase__

In [12]:
m_payout = 12 #withdraw every month (at the beginning of each month)
n_payout = 30 #30 years during retirement
cf_payout = 3000 #3,000 USD per month
FV_payout = 50000 #50,000 USD as final value of Annuity Contract
r_payout = 0.045 #4.5% on this investment for both Payout and Funding phase

#Payout Phase: คำนวณหา PV ของ Payout Phase แล้วนำ PV นี้ย้อนไปเป็น FV ของ Funding Phase และเนื่องจากโจทย์ให้ PV ของ Funding Phase มาแล้ว สามารถคำนวณหาจำนวนเดือนที่ต้อง fund ได้

In [13]:
PV_payout = npf.pv(rate = r_payout/m_payout, nper = n_payout * m_payout, pmt = cf_payout, fv = FV_payout, when = 'begin')
PV_payout
# Let FV_funding = -PV_payout (เครื่องหมายต้องตรงข้ามกัน)

-607298.5727512676

(result: -607298.5727512676)

__Funding Phase__

In [16]:
m_funding = 12 #deposit every month (at the end of each month)
cf_funding = -500 #deposit 500 USD, so negative sign
FV_funding = -PV_payout
PV_funding = 0 #No initial investment
r_funding = 0.045 #4.5% on this investment for both Payout and Funding phase

#find n

In [23]:
n = npf.nper(rate = r_funding / m_funding, pmt = cf_funding, pv = PV_funding, fv = FV_funding, when ='end')
n

array(458.09718839)

In [26]:
print("The funding phase will start in {} months".format(12 * 45 - np.ceil(n)))

print("retire in 45 years = 12 * 45 = 540 months (มีระยะเวลาทั้งหมด 450 เดือนก่อนเกษียณ อายุ เพราะฉะนั้นแล้วจากการคำนวณต้องใช้ทั้งหมด 459 เดือน (np.ceil(n) = 459) เพื่อให้ได้ FV_funding = 607298.5727512676 เพราะฉะนั้น สามารถเริ่มใส่เงินตอนเดือนที่ 81 ได้ เมื่อใส่เงินจนครบหรือผ่านไป 459 ก็จะได้ FV_funding = 607298.5727512676 ซึ่งเป็นเงินต้น PV_payout ในช่วง Payout Phase ต่อไป)")

The funding phase will start in 81.0 months
retire in 45 years = 12 * 45 = 540 months (มีระยะเวลาทั้งหมด 450 เดือนก่อนเกษียณ อายุ เพราะฉะนั้นแล้วจากการคำนวณต้องใช้ทั้งหมด 459 เดือน (np.ceil(n) = 459) เพื่อให้ได้ FV_funding = 607298.5727512676 เพราะฉะนั้น สามารถเริ่มใส่เงินตอนเดือนที่ 81 ได้ เมื่อใส่เงินจนครบหรือผ่านไป 459 ก็จะได้ FV_funding = 607298.5727512676 ซึ่งเป็นเงินต้น PV_payout ในช่วง Payout Phase ต่อไป)


(result: The funding phase will start in 81 months)

You consider to buy a new home for a Purchase Price of 700,000 USD. You decide to pay 300,000 USD cash. Therefore, you require another 400,000 USD from a Bank (Amortizing Mortgage) to realize the Deal. You agree with your local Bank on the following Mortgage terms:
- Fixed interest rate: 4.2% p.a. (compounded monthly)
- Loan Term with fixed interest rate is 35 years
- Outstanding Loan Amount after 35 Years shall be 0 USD
- Loan payments (interest & amortization) shall be made monthly in arrears 

5. Calculate the required __constant monthly payments__.

In [28]:
r = 0.042 #4.2 % per annual (compounded monthly)
n = 35
FV = 0
loan = 400000
m = 12

In [30]:
cf = npf.pmt(rate = r/m, nper = n*m, pv = loan, fv = FV)
cf

-1819.401880466396

(result: -1819.401880466396)

6. Calculate the __Repayment / Amortization Schedule__ on a monthly basis for the next 35 years. What is the amortization amount in the __very last period/month__?

In [31]:
#35 years * 12 months = 420 months
range(1, n*m+1) #1 ถึง 420

range(1, 421)

In [33]:
# Caculate Payment/Amortization Schedule in each period/month
amort = npf.ppmt(rate = r/m, per = range(1, n*m+1), nper = n*m, pv = loan, fv = FV)
amort

array([ -419.40188047,  -420.86978705,  -422.3428313 ,  -423.82103121,
        -425.30440482,  -426.79297024,  -428.28674563,  -429.78574924,
        -431.28999937,  -432.79951436,  -434.31431266,  -435.83441276,
        -437.3598332 ,  -438.89059262,  -440.42670969,  -441.96820318,
        -443.51509189,  -445.06739471,  -446.62513059,  -448.18831855,
        -449.75697766,  -451.33112709,  -452.91078603,  -454.49597378,
        -456.08670969,  -457.68301317,  -459.28490372,  -460.89240088,
        -462.50552429,  -464.12429362,  -465.74872865,  -467.3788492 ,
        -469.01467517,  -470.65622653,  -472.30352333,  -473.95658566,
        -475.61543371,  -477.28008773,  -478.95056803,  -480.62689502,
        -482.30908915,  -483.99717097,  -485.69116106,  -487.39108013,
        -489.09694891,  -490.80878823,  -492.52661899,  -494.25046216,
        -495.98033877,  -497.71626996,  -499.4582769 ,  -501.20638087,
        -502.96060321,  -504.72096532,  -506.4874887 ,  -508.26019491,
      

In [34]:
amort[-1] #last month

-1813.0561838230158

(result: -1813.0561838230158)

Company ABC evaluates two mutually exclusive Investment Projects (A and B) that lead to the following (estimated) increases in company cashflows/profits (in MUSD):

- Project A: [100, 110, 125, 150, 250, 500]
- Project B: [150, 200, 350, 200, 100, 100]

Both projects require an initial Investment of 800 Million USD (to buy Machinery and Plant).

The projects are mutually exclusive. Assume a company-wide required rate of return of 7.0% p.a. for comparable projects. 

7.  __Which project__ should ABC realize?

In [35]:
# run the cell!
cf_a = np.array([-800, 100, 110, 125, 150, 250, 500])
cf_b = np.array([-800, 150, 200, 350, 200, 100, 100])

In [36]:
r = 0.07 #Required rate of return 7% per annual

NPV_A = npf.npv(r, cf_a)
NPV_B = npf.npv(r, cf_b)

In [37]:
print("NPV A: {}, NPV B: {}".format(NPV_A, NPV_B))

NPV A: 117.42537723294546, NPV B: 91.09080119381395


(result: Project A)

In [38]:
print("Accept project A as NPV_A > NPV_B")

Accept project A as NPV_A > NPV_B


8. Calculate the (hypothetical) required rate of return where your decision would change (__crossover rate__).

In [39]:
npf.irr(cf_a - cf_b)
# hypothetical required rate of return can be caculated via the difference between the two cashflow and using irr built-in function

0.09549161767250536

(result: 0.09549161767250536)

# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Solutions (Stop here if you want to code on your own!)

# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

In [ ]:
# run the cell
import numpy as np
import numpy_financial as npf # NEW!!!

The ABC Company evaluates to buy an additional machine that will increase future profits/cashflows by
- 50 USD in t1, 
- 80 USD in t2,
- 120 USD in t3, 
- 150 USD in t4,
- 200 USD in t5. (each cf at period´s end)<br>

The machine costs 500 USD (Investment in to). ABC´s required rate of return (Cost of Capital) is 7.5% p.a. 

In [ ]:
# run the cell
cf = np.array([-500, 50, 80, 120, 150, 200])
cf

In [ ]:
# run the cell!
r = 0.075

1. Calculate the Net __Present Value (NPV)__ of the Project with __numpy financial functions__.

In [ ]:
npf.npv(r, cf)

(result: -36.03468883756554)

2. Calculate the __IRR__ of the Project with __numpy financial functions__.

In [ ]:
npf.irr(cf)

(result: 0.05234347404159134)

3. Should ABC __accept or reject__ the project?

reject

4. Assume a 20-year-old investor wants to retire in 45 years at the age of 65. He expects to earn 4.5% p.a. on his Investment during the Funding Phase and during the Payout Phase. He plans to start depositing 500 USD at the end of each month for n years in order to be able to withdraw 3,000 USD per month (at the beginning of each month) for 30 years during retirement. The Annuity Contract´s final Value shall be 50,000 USD when he is 95 years old (to cover some more years). __In how many months will the investor´s funding phase start__?

__Payout Phase__

In [ ]:
m = 12
cf_2 = 3000
n_2 = 30
FV_2 = 50000
r_2 = 0.045

In [ ]:
PV_2 = npf.pv(rate = r_2/m, nper = n_2*m, pmt = cf_2, fv = FV_2, when = "begin")
PV_2

(result: -607298.5727512676)

__Funding Phase__

In [ ]:
m = 12
cf_1 = -500
FV_1 = -PV_2
r_1 = 0.045
PV_1 = 0

In [ ]:
n = npf.nper(rate = r_1/m, pmt = cf_1, pv = PV_1, fv = FV_1)
n

In [ ]:
12 * 45 - np.ceil(n)

(result: The funding phase will start in 81 months)

You consider to buy a new home for a Purchase Price of 700,000 USD. You decide to pay 300,000 USD cash. Therefore, you require another 400,000 USD from a Bank (Amortizing Mortgage) to realize the Deal. You agree with your local Bank on the following Mortgage terms:
- Fixed interest rate: 4.2% p.a. (compounded monthly)
- Loan Term with fixed interest rate is 35 years
- Outstanding Loan Amount after 35 Years shall be 0 USD
- Loan payments (interest & amortization) shall be made monthly in arrears 

5. Calculate the required __constant monthly payments__.

In [ ]:
m = 12
loan = 400000
r = 0.042
fv = 0
n = 35

In [ ]:
cf = npf.pmt(rate = r/m, nper = n*m, pv = loan, fv = fv)
cf

(result: -1819.401880466396)

6. Calculate the __Repayment / Amortization Schedule__ on a monthly basis for the next 35 years. What is the amortization amount in the __very last period/month__?

In [ ]:
amort = npf.ppmt(rate = r/m, per = range(1, n*m + 1), nper = n*m, pv = loan, fv = fv)
amort

In [ ]:
amort[-1]

(result: -1813.0561838230158)

Company ABC evaluates two mutually exclusive Investment Projects (A and B) that lead to the following (estimated) increases in company cashflows/profits (in MUSD):

- Project A: [100, 110, 125, 150, 250, 500]
- Project B: [150, 200, 350, 200, 100, 100]

Both projects require an initial Investment of 800 Million USD (to buy Machinery and Plant).

The projects are mutually exclusive. Assume a company-wide required rate of return of 7.0% p.a. for comparable projects. 

7.  __Which project__ should ABC realize?

In [ ]:
# run the cell!
cf_a = np.array([-800, 100, 110, 125, 150, 250, 500])
cf_b = np.array([-800, 150, 200, 350, 200, 100, 100])

In [ ]:
npf.npv(0.07, cf_a)

In [ ]:
npf.npv(0.07, cf_b)

(result: Project A)

8. Calculate the (hypothetical) required rate of return where your decision would change (__crossover rate__).

In [ ]:
npf.irr(cf_a - cf_b)

(result: 0.09549161767250536)